### PR TITLE
Feature/1949 permanent subscriptions

### LINKED ISSUE #1949: Permanent subscriptions (it was "[BUG] in POST v2/subscriptions with Permanent subscription")
in POST v2/subscriptions with Permanent subscription.
> expires: Subscription expiration date in ISO8601 format. Permanent subscriptions must omit this field.

#### Subcription request
```
POST http://localhost:1026/v2/subscriptions
Content-Type: application/json
Fiware-Service: test_casub_happy_path
Fiware-ServicePath: /test
 {
 	"description": "my first subscription",
 	"subject": {
 		"entities": [{
 			"idPattern": ".*",
 			"type": "room"
 		}],
 		"condition": {
 			"attributes": ["temperature_0", "temperature_1", "temperature_2"],
 			"expression": {
 				"q": "temperature>40"
 			}
 		},
 		"notification": {
 			"callback": "http://localhost:1234",
 			"throttling": 5,
 			"attributes": ["temperature_0", "temperature_1", "temperature_2"]
 		}
 	}
 }
```

#### subscription response
```
http code: 400
date: Thu, 24 Mar 2016 10:21:14 GMT
connection: Keep-Alive
content-type: application/json
content-length: 79
{"error":"BadRequest","description":"no expiration for subscription specified"}
```

#### Expected response
```
http code: 201
```

### PR DESCRIPTION
**Pull request by crbrox**
@fgalan 
@kzangeli 

Add permanent subscriptions. 
Close #1949

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -12,3 +12,4 @@
 - Fix: POST /v2/subscriptions, attributes in notification are optional (#1952)
 - Add: Correlator as HTTP header and in log file (Issue #1916)
 - Fix: no longer considering service (tenant) for subscriptions when broker is started without '-multitenant' CLI option (Issue #1898)
+- Add: Add permanent subscriptions (Issue #1949)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: The following piece of documentation needs to be fixed (https://fiware-orion.readthedocs.org/en/develop/admin/database_model/index.html#csubs-collection):

> expiration: this is the timestamp on which the subscription expires. This is calculated using the duration parameter included in the subscribeContext operation (basically, sum "now" and duration) and will be recalculated when an updateContextSubscription is received (see programmers guide). For permanent subscriptions (allowed in NGSIv2) it is set to -1.

Maybe chaning 

For permanent subscriptions (allowed in NGSIv2) it is set to -1.

by

For permanent subscriptions (allowed in NGSIv2) an absurdely high value is used (see PERMANENT_SUBS_DATETIME in the source code)

File: test/functionalTest/cases/1949_permanent_subscriptions/1949_permanent_subscriptions.test [hunk 2]
```diff
@@ -0,0 +1,240 @@
+# Copyright 2016 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/1949_permanent_subscriptions/1949_permanent_subscriptions.test]: I would suggest to add a test to check that update/delete expiration works, I mean:

1. Create regular subscription (with expires)
2. Get subscription to see the 'expires' field
3. Update subscription to make it permanent
4. Get subscription to check that 'exipres' field is not shown
5. Update subscription to provide a expiration date again
6. Get subscription to see the 'expires' field

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (crbrox) [inline CHANGES_NEXT_RELEASE]: Fixed 42cacac

File: test/functionalTest/cases/1949_permanent_subscriptions/1949_permanent_subscriptions.test (hunk 2 above)
Comment (crbrox) [inline test/functionalTest/cases/1949_permanent_subscriptions/1949_permanent_subscriptions.test]: Fixed 1c1daec

**Comment by kzangeli** [comment]
LGTM

**Comment by fgalan** [comment]
LGTM

**Pull request merged**